samples/sensor/bme280/src/main.c

```diff
@@ -9,25 +10,28 @@
 #include <devicetree.h>
 #include <drivers/sensor.h>
 
-#define BME280 DT_INST(0, bosch_bme280)
-
-#if DT_NODE_HAS_STATUS(BME280, okay)
-#define BME280_LABEL DT_LABEL(BME280)
-#else
-#error Your devicetree has no enabled nodes with compatible "bosch,bme280"
-#define BME280_LABEL "<none>"
-#endif
+/*
+ * Devicetree node identifier for the BME280 device we should use.
+ * This is configured via the "sample-bme280" property in /aliases.
+ * See app.overlay for more information.
+ */
+#define BME280_NODE DT_ALIAS(sample_bme280)
```

Comment:
This breaks on boards that have a BME280, and don't provide a sample alias for it.  IMO samples that demonstrate functionality of a device should "just work" on platforms that have that device, and we shouldn't have to add aliases to the board DTS to make that happen.

Can that be fixed by doing an existence check on the `sample_bme280` alias, and if it isn't there looking for the first instance of the `bosch,bme280` device?

Reply:
> Can that be fixed by doing an existence check on the `sample_bme280` alias, and if it isn't there looking for the first instance of the `bosch,bme280` device?

I'm trying to start a trend towards avoid using instance numbers in samples. I think we should try to avoid them because explaining the semantics is quite complicated.

Reply:
Also responding to https://github.com/zephyrproject-rtos/zephyr/pull/31056#discussion_r556275072:

I see the value in that, but IMO it shouldn't come at the cost of making samples work with boards that already have the sensor only if those boards add special aliases.  It weakens the rationale for having access by devicetree path or compatible instance.  Maybe we could have a special set of macros for detecting and using the first instance of a compatible that doesn't expose the whole instance infrastructure.

Reply:
> Can that be fixed by doing an existence check on the sample_bme280 alias, and if it isn't there looking for the first instance of the bosch,bme280 device?

There shouldn't be a difference between dts  and dts + overlay from application point of view. dts resolution happens before compilation comes into play and from that regard, this should be transparent.
I think also that alias should not be needed (at all then), but I admit I can't challenge with a counter-proposal.

Reply:
> Maybe we could have a special set of macros for detecting and using the first instance of a compatible that doesn't expose the whole instance infrastructure. That would handle the needs of samples in the vast majority of cases.

I thought about this a bit and am proposing a new macro, `DEVICE_DT_GET_ANY(compat)`, that does exactly that. I've moved the last patch to using that instead of aliases. This required splitting up the overlays into separate ones for arduino SPI and arduino I2C, but should allow boards like the ruuvitag to work out of the box.

I like this solution better and appreciate the pushback; @pabigot.